# Random Forest Classifier Tutorial
In this notebook, we'll walk through a practical example of using the Random Forest Classifier to predict wine quality using the Kaggle dataset.

In [56]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier, export_graphviz
from sklearn.ensemble import RandomForestClassifier
from sklearn import tree
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, confusion_matrix, roc_curve, roc_auc_score


In [58]:
data = pd.read_csv("winequalityN.csv")
data.head()


type  fixed acidity  volatile acidity  citric acid  residual sugar  \
0  white            7.0              0.27         0.36            20.7   
1  white            6.3              0.30         0.34             1.6   
2  white            8.1              0.28         0.40             6.9   
3  white            7.2              0.23         0.32             8.5   
4  white            7.2              0.23         0.32             8.5   

   chlorides  free sulfur dioxide  total sulfur dioxide  density    pH  \
0      0.045                 45.0                 170.0   1.0010  3.00   
1      0.049                 14.0                 132.0   0.9940  3.30   
2      0.050                 30.0                  97.0   0.9951  3.26   
3      0.058                 47.0                 186.0   0.9956  3.19   
4      0.058                 47.0                 186.0   0.9956  3.19   

   sulphates  alcohol  quality  
0       0.45      8.8        6  
1       0.49      9.5        6  
2       0.44     10.1        6  
3       0.40      9.9        6  
4       0.40      9.9        6

In [60]:
data = data.dropna()
data.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    6463.000000       6463.000000  6463.000000     6463.000000   
mean        7.217755          0.339589     0.318758        5.443958   
std         1.297913          0.164639     0.145252        4.756852   
min         3.800000          0.080000     0.000000        0.600000   
25%         6.400000          0.230000     0.250000        1.800000   
50%         7.000000          0.290000     0.310000        3.000000   
75%         7.700000          0.400000     0.390000        8.100000   
max        15.900000          1.580000     1.660000       65.800000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  6463.000000          6463.000000           6463.000000  6463.000000   
mean      0.056056            30.516865            115.694492     0.994698   
std       0.035076            17.758815             56.526736     0.003001   
min       0.009000             1.000000              6.000000     0.987110   
25%       0.038000            17.000000             77.000000     0.992330   
50%       0.047000            29.000000            118.000000     0.994890   
75%       0.065000            41.000000            156.000000     0.997000   
max       0.611000           289.000000            440.000000     1.038980   

                pH    sulphates      alcohol      quality  
count  6463.000000  6463.000000  6463.000000  6463.000000  
mean      3.218332     0.531150    10.492825     5.818505  
std       0.160650     0.148913     1.193128     0.873286  
min       2.720000     0.220000     8.000000     3.000000  
25%       3.110000     0.430000     9.500000     5.000000  
50%       3.210000     0.510000    10.300000     6.000000  
75%       3.320000     0.600000    11.300000     6.000000  
max       4.010000     2.000000    14.900000     9.000000

The dataset consists of the following Input variables:
1 - fixed acidity
2 - volatile acidity
3 - citric acid
4 - residual sugar
5 - chlorides
6 - free sulfur dioxide
7 - total sulfur dioxide
8 - density
9 - pH
10 - sulphates
11 - alcohol

And the Output variable gives the quality of the wine based on the input variables:
12 - quality (score between 0 and 10)

## Train Random Forest Model

In [64]:
# Encode the 'type' column (e.g., 'red', 'white') into numbers
label_encoder = LabelEncoder()
data['type'] = label_encoder.fit_transform(data['type'])

X = data.drop(columns='quality')
y = data['quality']


x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=355)


In [89]:
rand_clf = RandomForestClassifier(random_state=6)
rand_clf.fit(x_train, y_train)  # Fitting the model to the training data


RandomForestClassifier(random_state=6)

## Accuracy for Random forest classifer

In [83]:
rand_clf.score(x_test, y_test)

0.6740587931923672

### Accuracy for Random Forest Classifier is 0.6740587931923672 

# Decision Tree Classifer

In [105]:
dt_clf = DecisionTreeClassifier(criterion='entropy', max_depth=10, random_state=42)
dt_clf.fit(x_train, y_train)

DecisionTreeClassifier(criterion='entropy', max_depth=10, random_state=42)

In [107]:
dt_clf.score(x_test, y_test)

0.5394533264569366

### Accuracy for Decision Tree Classifer is 0.5394533264569366

 # HENCE ACCURACY SCORE FOR RANDOM FOREST CLASSIFIER > DECISION TREE CLASSIFIER